src/codecs/avif/decoder.rs

```diff
@@ -0,0 +1,156 @@
+//! Decoding of AVIF images.
+///
+/// The [AVIF] specification defines an image derivative of the AV1 bitstream, an open video codec.
+///
+/// [AVIF]: https://aomediacodec.github.io/av1-avif/
+
+use std::convert::TryFrom;
+use std::io::{self, Cursor, Read};
+use std::marker::PhantomData;
+use std::mem;
+
+use crate::{ColorType, ImageResult, ImageDecoder};
+
+use mp4parse::read_avif;
+use dav1d::{PlanarImageComponent, PixelLayout};
+
+/// AVIF Decoder.
+///
+/// Reads one image into the chosen input.
+pub struct AvifDecoder<R> {
+    inner: PhantomData<R>,
+    picture: dav1d::Picture,
+    alpha_picture: Option<dav1d::Picture>,
+}
+
+impl<R: Read> AvifDecoder<R> {
+    /// Create a new decoder that reads its input from `r`.
+    pub fn new(mut r: R) -> ImageResult<Self> {
+        let ctx = read_avif(&mut r).unwrap();
+        let mut primary_decoder = dav1d::Decoder::new();
+        primary_decoder.send_data(ctx.primary_item(), None, None, None).unwrap();
+        let picture = primary_decoder.get_picture().unwrap();
+        let alpha_picture = if let Some(alpha_item) = ctx.alpha_item() {
+            let mut alpha_decoder = dav1d::Decoder::new();
+            alpha_decoder.send_data(alpha_item, None, None, None).unwrap();
+            Some(alpha_decoder.get_picture().unwrap())
+        } else {
+            None
+        };
+        assert_eq!(picture.bit_depth(), 8);
+        Ok(AvifDecoder {
+            inner: PhantomData,
+            picture,
+            alpha_picture,
+        })
+    }
+}
+
+/// Wrapper struct around a `Cursor<Vec<u8>>`
+pub struct AvifReader<R>(Cursor<Vec<u8>>, PhantomData<R>);
+impl<R> Read for AvifReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.0.read(buf)
+    }
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        if self.0.position() == 0 && buf.is_empty() {
+            mem::swap(buf, self.0.get_mut());
+            Ok(buf.len())
+        } else {
+            self.0.read_to_end(buf)
+        }
+    }
+}
+
+impl<'a, R: 'a + Read> ImageDecoder<'a> for AvifDecoder<R> {
+    type Reader = AvifReader<R>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        (self.picture.width(), self.picture.height())
+    }
+
+    fn color_type(&self) -> ColorType {
+        ColorType::Bgra8
+    }
+
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+        let data = self.picture.plane_data_ptr(PlanarImageComponent::Y) as *mut _;
+        let (width, height) = self.picture.plane_data_geometry(PlanarImageComponent::Y);
+        let vec = unsafe { Vec::from_raw_parts(data, (width * height) as usize, (width * height) as usize) };
```

Comment:
`Vec::from_raw_parts` is only safe when `data` was allocated with the Rust global allocator, and exactly with the size given in `capacity` and allign of its data type, that is 1. There is no proof of these facts, so allocate this into a new vector for exampling using `[u8]::to_vec`.

Reply:
I completely missed the `.plane()` API, which implements `AsRef<[u8]>` and is thus perfectly safe to use here. The latest version now removed all `unsafe` from this module.